components/ContentEditor/ContentEditor.tsx

```diff
@@ -0,0 +1,184 @@
+"use client";
+
+import { useRef, useState } from "react";
+import { Editor } from "@tiptap/core";
+import { useEditor, EditorContent, BubbleMenu } from "@tiptap/react";
+import StarterKit from "@tiptap/starter-kit";
+
+import LoadingDots from "../LoadingDots";
+import {
+  GrammarChecker,
+  GrammarCheckerOperations,
+} from "./extensions/GrammarChecker";
+import { Match, Range } from "./extensions/GrammarChecker.types";
+
+const defaultContent = `<p>Biology is a really unique scient to study. There are alott of different aspects to it, such as ecology, genetics, and physiology. One of the most interesitng things to learn about in biology is animals and the way they behave. For example, did you know that some birds give hugs to their babies to keep them warm? That's so cute!</p>
+
+<p>Another important aspect of biology is understanding the structure and function of different living things. Cells are the basic building blocks of all living organisms, and they are responsible for carrying out all of the processes necessary for life. Studying the biology of cells is important for understanding everything from how the body works to how diseases develop.</p>`;
+
+const ContentEditor = () => {
+  const [isLoading, setIsLoading] = useState(false);
+
+  const editor = useEditor({
+    extensions: [
+      StarterKit,
+      GrammarChecker.configure({
+        documentId: "grammar-checker-v1",
+      }),
+    ],
+    content: defaultContent,
+    editorProps: {
+      attributes: {
+        class:
+          "border border-gray-300 min-h-[200px] p-3 prose dark:prose-invert focus:outline-none max-w-full",
+      },
+    },
+    onUpdate({ editor }) {
+      setTimeout(() => updateMatch(editor as any));
+    },
+    onSelectionUpdate({ editor }) {
+      setTimeout(() => updateMatch(editor as any));
+    },
+    onTransaction({ transaction: tr }) {
+      if (tr.getMeta(GrammarCheckerOperations.LoadingTransactionName))
+        loading.current = true;
+      else loading.current = false;
+    },
+  });
+
+  const checkGrammar = async () => {
+    const cleanText = editor?.getText();
+
+    if (!cleanText) {
+      console.log("content is missing: ", cleanText);
+      alert("Content is missing!");
+      return;
+    }
+
+    // max word limit
+    if (cleanText.split(" ")?.length > 800) {
+      alert("You may only use 800 words at once!");
+      return;
+    }
+
+    try {
+      setIsLoading(true);
+      const response = await fetch("/api/grammar", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({ content: cleanText }),
+      });
+
+      if (!response.ok) {
+        const errorResponse = await response.json();
+        alert(errorResponse?.error);
+      } else {
+        const data = await response.json();
+
+        if (data.result.corrections.length >= 0) {
+          editor?.commands.proofread(data.result.corrections);
```

Comment:
Where is this proofread command coming from? Is this part of the editor or written by you?

Reply:
I found it